### ISSUE TITLE
NullPointerException: org.jgrapht.io.DOTImporter$State.getId

### ISSUE THREAD

**Issue by spi-x-i** [jgrapht/jgrapht#782]
```
 * JGraphT version: 1.3.0
 * Java version (java -version)/platform:  1.8
```

**Issue**
When I try to import my DOT graph I got a `NullPointerException`

```
java.lang.NullPointerException:
[info]   at org.jgrapht.io.DOTImporter$State.getId(DOTImporter.java:730)
[info]   at org.jgrapht.io.DOTImporter$CreateGraphDOTListener.exitNodeStatement(DOTImporter.java:479)
[info]   at org.jgrapht.io.DOTParser$NodeStatementContext.exitRule(DOTParser.java:778)
[info]   at org.antlr.v4.runtime.Parser.triggerExitRuleEvent(Parser.java:408)
[info]   at org.antlr.v4.runtime.Parser.exitRule(Parser.java:642)
[info]   at org.jgrapht.io.DOTParser.statement(DOTParser.java:398)
[info]   at org.jgrapht.io.DOTParser.compoundStatement(DOTParser.java:176)
[info]   at org.jgrapht.io.DOTParser.graph(DOTParser.java:125)
[info]   at org.jgrapht.io.DOTImporter.importGraph(DOTImporter.java:142)
```

**Steps to reproduce (small coding example)**
This is the graph I need to import
```
strict digraph G_testdefinition_G {

  0 [ label="Rounding" 1e916492-6942-4988-a0db-0e4083a5ce90="{"type":"record","name":"operator1OutputGroup1Schema","namespace":"my.namespace","fields":[]}" name="RoundFields" presentationName="operator1" action="transformation" namespace="my.namespace.operators.transformation" category="transformation" style="{"top":"0px","left":"0px"}" mode="default" inputs="[{"group":"inputGroup1","accept":"*"}]" outputs="[{"group":"outputGroup1","produce":"mirrored","meta":{"mirror":"inputGroup1"}}]" properties="{"default":[{"group":"default","arity":"single","name":"digits","type":"text","label":"Number of Digits to Retain","required":true,"description":"The number of decimal digits to be preserve along rounding.","values":{"from":"given"}},{"group":"default","arity":"single","name":"roundMode","type":"select","label":"Rounding Strategy","required":true,"description":"Rounding either to lower (Floor) or higher (Ceiling) multiple. Default is Ceiling","values":{"from":"given","content":{"labels":["Ceiling","Floor"],"names":["ceiling","floor"]}}},{"group":"default","arity":"multiple","name":"impliedFields","type":"select","label":"Fields to Round Off","required":false,"description":"Fields to be rounded off in output","values":{"from":"input","content":{"group":["inputGroup1"]}}}]}" data="{}" inputFormatUUID="{}" outputFormatUUID="{"outputGroup1":"1e916492-6942-4988-a0db-0e4083a5ce90"}" errors="[]" ];

  1 [ label="Rounding" 9d17c6ce-bdf8-46cb-aab2-2bb8fbeaded7="{"type":"record","name":"operator2OutputGroup1Schema","namespace":"my.namespace","fields":[]}" name="RoundFields" presentationName="operator2" action="transformation" namespace="my.namespace.operators.transformation" category="transformation" style="{"top":"0px","left":"0px"}" mode="default" inputs="[{"group":"inputGroup1","accept":"*"}]" outputs="[{"group":"outputGroup1","produce":"mirrored","meta":{"mirror":"inputGroup1"}}]" properties="{"default":[{"group":"default","arity":"single","name":"digits","type":"text","label":"Number of Digits to Retain","required":true,"description":"The number of decimal digits to be preserve along rounding.","values":{"from":"given"}},{"group":"default","arity":"single","name":"roundMode","type":"select","label":"Rounding Strategy","required":true,"description":"Rounding either to lower (Floor) or higher (Ceiling) multiple. Default is Ceiling","values":{"from":"given","content":{"labels":["Ceiling","Floor"],"names":["ceiling","floor"]}}},{"group":"default","arity":"multiple","name":"impliedFields","type":"select","label":"Fields to Round Off","required":false,"description":"Fields to be rounded off in output","values":{"from":"input","content":{"group":["inputGroup1"]}}}]}" data="{}" inputFormatUUID="{}" outputFormatUUID="{"outputGroup1":"9d17c6ce-bdf8-46cb-aab2-2bb8fbeaded7"}" errors="[]" ];

  0 -> 1 [ label="0" ];
}
```

**Other information** 
Using scala 2.11.12. I tried also `jgrapht-io-1.3.1-SNAPSHOT` version because of [this](https://github.com/jgrapht/jgrapht/issues/746) but I encounter the same error.  Thank you for your help.

**Comment by d-michail** [comment]
Hi, 

I tried your input in Java and the parser complains. You are probably seeing a null pointer due to the fact that you are using Scala. 

Υour input is not valid. See here for the grammar (https://www.graphviz.org/doc/info/lang.html). The rules for identifiers are 

```
An ID is one of the following:
  - Any string of alphabetic ([a-zA-Z\200-\377]) characters, underscores ('_') or digits ([0-9]), not beginning with a digit;
  - a numeral [-]?(.[0-9]+ | [0-9]+(.[0-9]*)? );
  - any double-quoted string ("...") possibly containing escaped quotes (\")1;
  - an HTML string (<...>).
```
which means that for the very least you must add double-quotes around the uids that you are using. 

However, your input has more problems than that. Your double-quotes escaping is malformed. 

After a quick look, the following which is part of your input is not valid 
```
style="{"top":"0px","left":"0px"}"
```
As a general guideline you cannot use a language (json) inside another language (dot) without proper escaping. Your input should have been 

```
style="{\"top\":\"0px\",\"left\":\"0px\"}"
```

When submitting bug reports in the feature, try constructing a minimal reproducible example (https://stackoverflow.com/help/minimal-reproducible-example).

**Comment by spi-x-i** [comment]
Hi, thank you for your timely answer and sorry if my question was actually bad formulated: I posted the unescaped version of the graph, my actual version is escaped.

The issue was the **UUID** incompatibility as a key as you were guessing. Even introducing double quotes was not enough. I needed to remove dashes and prefix them with underscores as
```7270d88b-9c5d-45c9-aa5d-9b38aaaf0f81 --> _7270d88b9c5d45c9aa5d9b38aaaf0f81``` .

This made the above UUID a valid key string.
I think the error comes really hard to understand in Scala and I don't know if we can keep it better somehow, anyway the issue is solved by my side. Thank you!
